src/boot/efi/boot.c

```diff
@@ -1624,12 +1633,36 @@ static INTN config_entry_compare(const ConfigEntry *a, const ConfigEntry *b) {
         assert(a);
         assert(b);
 
-        /* Order entries that have no tries left to the beginning of the list */
-        if (a->tries_left != 0 && b->tries_left == 0)
-                return 1;
+        /* Order entries that have no tries left to the end of the list */
         if (a->tries_left == 0 && b->tries_left != 0)
+                return 1;
+        if (a->tries_left != 0 && b->tries_left == 0)
                 return -1;
 
+        /* First, order by sort key (alphabetically upwards), if it is defined. */
+        if (a->sort_key && b->sort_key) {
+                r = strcmp(a->sort_key, b->sort_key);
+                if (r != 0)
+                        return r;
+
+                /* If multiple installations of the same OS are around, group by machine ID */
+                if (a->machine_id && b->machine_id) {
+                        r = strcmp(a->machine_id, b->machine_id);
+                        if (r != 0)
+                                return r;
+                }
+
+                /* If the sort key was defined, then order by version now (downwards, putting the newest first) */
+                if (a->version && b->version) {
```

Comment:
`strverscmp_improved` is NULL safe, so this can be dropped.

Reply:
well, not really. if one of the entries has version set and the other does not we don't want to compare by version. if we'd blindly call strverscmp_improved() it would return != in that case however. 

or in other words: if you compare one entry with and one without version we shouldn't use the version field at all, instead of ordering the one without the version before the one with the version

Reply:
you are probably right, we should group by machine ID early on.

Reply:
hmm, actually i don't follow. we sort by machined *before* version, precisely to have this order:

1. machine1 version 3
2. machine1 version 2
3. machine1 version 1
4. machine2 version 3
5. machine2 version 2
6. machine2 version 1

I'd argue that's exactly what we want.

Reply:
You're right. I think the diff output just confused me